### DISCUSSION TITLE
[Discussion] typescript-rtti vs tst-reflect

### DISCUSSION THREAD

**Discussion by Hookyns** [typescript-rtti/typescript-rtti#18] [General]
Hello,

I've found your project in https://github.com/microsoft/TypeScript/issues/3628 and I've got an idea (regarding to the discussion about making a standard).. would you like to join our project? I think our projects are doing the same thing but our project is much further IMHO. 🙏

**Comment by rezonant** [comment]
You didn't link to your project :-)

**Comment by Hookyns** [comment]
Sorry 🤦‍♂️ , https://github.com/Hookyns/ts-reflection

**Comment by rezonant** [comment]
Initial thoughts:
- Why not use `Type.getType(value)` instead of `getType()` -- you've got this nice class to hang it on, why not use it?
- I don't see where you have an example of calling `getType()` to retrieve a `Type` in the README. Is it getting the type _of the value_ or the type that Typescript sees it at compile time (ie `<T>`)? Because the second one is not useful- I should be able to introspect a value that is ostensibly unknown at compile-time to discover its type at runtime.
- With the metadata being stored in one place (`metadata.lib.js`) how does this work with multiple packages?
    * If I requested type metadata for a type that comes from an imported NPM package, does it fetch it from the imported package's `metadata.lib.js` or emit another (duplicate) type in the calling package?

**Comment by rezonant** [comment]
Wow I missed the simple example bit 

```
    const type = getType<TType>();
```

See, this is not runtime type metadata for _values_, it is runtime type metadata for _types_. This is a fundamentally different operation unless I can acquire the `Type` value from the object's constructor in some way (ie `value.constructor.type` or something)

**Comment by rezonant** [comment]
I believe the library you've built satisfies a different need than typescript-rtti. I do think we all need to come together for a common solution, and maybe these two projects are two sides of a similar coin, but the use cases are not the same

**Comment by Hookyns** [comment]
From your Usage:

```typescript
let aClass = new ReflectedClass(A);
console.log(aClass.parameterNames); // ["someValue", "someOtherValue"]
console.log(aClass.parameters[0].name); // "someValue"
console.log(aClass.getParameter('someValue').type); // Number
console.log(aClass.getParameter('someOtherValue').type); // String

let bClass = new ReflectedClass(B);
console.log(bClass.propertyNames) // ["foo", "bar"]
console.log(bClass.getProperty('foo').type) // A
console.log(bClass.getProperty('foo').visibility) // "private"
console.log(bClass.getProperty('bar').type) // Number
console.log(bClass.methodNames) // [baz]
console.log(bClass.getMethod('baz').returnType) // A
```

Equivalent in `tst-reflect`:

```typescript
import { getType } from "tst-reflect";

export class A {
	constructor(
		readonly someValue : number,
		private someOtherValue : string
	) {
	}
}

export class B {
	private foo : A;
	bar = 123;
	baz(): A {
		return this.foo;
	}
}

let aClass = getType<A>();

const parameters = aClass.getConstructors()[0].getParameters();
console.log(parameters.map(param => param.name)); // ["someValue", "someOtherValue"]
console.log(parameters[0].name); // "someValue"
console.log(parameters.find(param => param.name == 'someValue').type.name); // Number
console.log(parameters.find(param => param.name == 'someOtherValue').type.name); // String

let bClass = getType<B>();
const properties = bClass.getProperties();
const methods = bClass.getMethods();
console.log(properties.map(prop => prop.name)) // ["foo", "bar"]
console.log(properties.find(prop => prop.name == 'foo').type.name) // A
console.log(properties.find(prop => prop.name == 'foo').accessModifier) // AccessModifier.Private
console.log(properties.find(prop => prop.name == 'bar').type.name) // Number
console.log(methods.map(method => method.name)) // [baz]
console.log(methods.find(method => method.name == 'baz').returnType.name) // A
```

**Comment by Hookyns** [comment]
Even this works

```typescript
const b = Reflect.construct(bClass.ctor, [0, "something"]);
console.log(b instanceof B); // true
```


>From your Usage:
> ...
> Equivalent in tst-reflect:
> ...

The same thing is possible with interfaces, abstract classes and "type" declarations.

**Comment by rezonant** [comment]
Continuing from https://github.com/microsoft/TypeScript/issues/3628#issuecomment-1030416371

**Comment by rezonant** [comment]
Oh youve got it my bad

**Comment by rezonant** [comment]
So doing what tst-reflect does in typescript-rtti is possible but the reverse isn't. I might have even looked at your project before starting this- but getType<T>() wouldn't have been sufficient for the use cases here

```
let type = new ReflectedClass(A); // vs getType<A>();
```

```
let type = new ReflectedClass(arbitraryValue.constructor); // vs getType<??>()
```

**Comment by Hookyns** [comment]
Now I understand how you do that.. it is big difference..

I had plan to implement the same thing,.. to allow calls `getType(variable)` which would return instance of Type too but created from runtime information. (We have some discussion somewhere in our repository on this topic)

But I rejected that, because those Type would never contain such detailed info (import info (paths), fullNames - unique identifiers) so they cannot be compatible and check like `Type.is(Type)` will fail.

**Comment by rezonant** [comment]
I would say import graphs and file names (of where the type originated) are outside the scope of typescript-rtti so that wasn't an issue here

What are the use cases for that?

**Comment by Hookyns** [comment]
Well you cannot do this on interfaces, can you?

```typescript
interface B {
	foo : string;
	bar: number;
	baz(): B;
}

let bClass = getType<B>();
const properties = bClass.getProperties();
const methods = bClass.getMethods();
console.log(properties.map(prop => prop.name)) // ["foo", "bar"]
console.log(properties.find(prop => prop.name == 'foo').type.name) // string
console.log(properties.find(prop => prop.name == 'bar').type.name) // number
console.log(methods.map(method => method.name)) // [baz]
console.log(methods.find(method => method.name == 'baz').returnType.name) // B
```

**Comment by Hookyns** [comment]
> I would say import graphs and file names (of where the type originated) are outside the scope of typescript-rtti so that wasn't an issue here
> 
> What are the use cases for that?

Information about modules is good for Node.js framework. That's quite useless in browser but still, u can use it in browser anyway.

```typescript
const someType = getType<SomeType>(); 
// You can get this reference as parameter or you have just `import type { SomeType} from "./SomeType";` or ,.. idk.. 
// So you don't have SomeType imported and you don't have the reference. You have only Type instance.

// --------

function doSomething(type: Type) {
    const instance = new type.ctor(); // This contains static `require("./some/module.js")` so bundlers proccess this...

    instance.doSomethingWithThisInstance();
}
```

Or some Node.js framework usage:
```typescript
const controllers = Type.getTypes()
    .filter(type => type.decorators.some(dec => dec.name == "Controller"))
    .filter(controller => controller.getMethods().some(method => method.name == "getList"))
);

controllers
    .map(controller => Reflect.construct(controller.ctor, []))
    .forEach(controllerInstance => controllerInstance.getList());

```

**Comment by rezonant** [comment]
The first example works in this project of course, you need the constructor to build a Reflected class.

The second one though we have no way to do right now, but one could imagine it could be added if we adopted a central store like what tst-reflect has. But this could be dangerous--- imagine this:

delete-all.ts
```
class SomeUnrelatedClass {}
deleteAllStuff()
```

Here deleteAllStuff() runs when your code constructs SomeUnrelatedClass

**Comment by Hookyns** [comment]
Or reallife example https://github.com/Hookyns/NetLeaf/tree/main/libraries/extensions-dependency-injection-typed

**Comment by rezonant** [comment]
> Well you cannot do this on interfaces, can you?
> 
> ```ts
> interface B {
> 	foo : string;
> 	bar: number;
> 	baz(): B;
> }
> 
> let bClass = getType<B>();
> const properties = bClass.getProperties();
> const methods = bClass.getMethods();
> console.log(properties.map(prop => prop.name)) // ["foo", "bar"]
> console.log(properties.find(prop => prop.name == 'foo').type.name) // string
> console.log(properties.find(prop => prop.name == 'bar').type.name) // number
> console.log(methods.map(method => method.name)) // [baz]
> console.log(methods.find(method => method.name == 'baz').returnType.name) // B
> ```

Not in typescript-rtti _yet_ though I do have some code that would do something similar to `getType<T>()` in a separate branch-- I'm not sure yet whether it will be mainlined.

**Comment by Hookyns** [comment]
> Here deleteAllStuff() runs when your code constructs SomeUnrelatedClass

**This is good point!** 

_A lot of frameworks do the same thing the hard way,.. traversing FS, looking for files in specified directories and dynamically import them to execute actions in controllers._

We are working on config options `include` and `exclude` which are glob patterns so you can exclude `node_modules/**/*` to prevent proccesing of unknown files (even if you use those types in your code).
Maybe I can add option `onlySafeImports: boolean`. Types will be created but there will be no way to import them. 👍

**Comment by rezonant** [comment]
So the idea is "wouldn't it be nice if I didn't need this?" right?

```typescript
// library.ts

export let CONTROLLERS = [];

function registerController(klass) {
    CONTROLLERS.push(klass);
}

export function Controller() {
    return (target) => registerController(target);
}
```

```typescript
// some-controller.ts
@Controller()
class SomeController {
}
```

**Comment by rezonant** [comment]
I'm just not sure we should go against how imports in Javascript _work_-- that if something isn't imported it doesn't run. It has ramifications that are not easy to anticipate

**Comment by Hookyns** [comment]
Well type.ctor invokes that import (it's a getter) so by using `ctor` you are telling "Import that module for me".
Anyway this is mainly for authors of frameworks and libs,..

I've made MVC framework in Node.js and I was loading modules from `controllers` folder (simplified) and registering controllers and their actions into the router (after that I cleared require cache). It's just automation,.. it is annoying to make thousands of imports if you don't have to.
My new contributor make MVC framework too and he does very similar thing. You can have decorators everywhere but without imports u have no chance to get that information so you don't know you have something decorated.. that's what reflection is supposed to do...

It is the same how reflection in C# works. You can do all of that. You don't have to import anything,.. reflection do the job.

**Comment by rezonant** [comment]
> You can have decorators everywhere but without imports u have no chance to get that information so you don't know you have something decorated.. that's what reflection is supposed to do...

So [Alterior](https://github.com/alterior-mvc/alterior) [I am the author] used to work this way in version 2.x. It would do what I had above where there was registration. But following Angular's lead we now declare these as part of a "module" that gets exposed to the application framework instead, and it's worked great. I've built a handful of serious applications this way and it's been just fine.

So that's like this:

```typescript
import { SomeController } from './some-controller';
import { Module } from '@alterior/runtime`;

@Module({
    controllers: [
        SomeController
    ]
})
class MyModule {

}
```

Since then with a move toward building isomorphism by convention it has evolved to something more like:

```typescript
import { UserController } from './user.controller';
import { WebService } from '@alterior/web-server`;

@WebService()
export class MyService {
    @Get('/foo') async foo() {
        return { status: 'success' };
    }

    @Mount() users : UserController;
}

Application.bootstrap(MyService);
```

And I have to say this is a great way to do MVC. All controllers are "mounted" declaratively and it eliminates the import problem, provides a good developer abstraction for how this works, **and** allows multiple distinct web services to live in one codebase without any need for global enumeration of controllers.

Naturally you can do this deeply-- have a `@Controller()` that `@Mount()`s other controllers.

From an isomorphism point of view, the goal is that the framework will emit a client definition that is "shaped" exactly like the server you built but with none of the server code, only the "service"/"controller" classes you've built. So in _theory_ a client could reuse this:

```typescript
import { MyService } from `@.../my-service`;

let service = new MyService("https://my-location.example.com/");
await service.foo();  // returns you back `{status: 'success'}`
await service.users.list(); // etc
```

**Comment by rezonant** [comment]
Of course none of Alterior requires anything other than (A) `emitDecoratorMetadata: true` and (B) `import 'reflect-metadata';` at the root of the project, and for some features (C) `import 'zone.js';` because Zones should be standard and I wish more people understood why :-\

**Comment by rezonant** [comment]
@Hookyns Does `Type.getTypes()` return classes* which are not exported? If so, is `ctor` just undefined?

**Comment by Hookyns** [comment]
> because Zones should be standard and I wish more people understood why

There are a lot of things that people don't know or don't use because they don't know what is it good for. Among other languages, I also work with C# and it has many nice features which nobody miss in JavaScript/Node.js because they don't know that something like that is even possible so they don't know they want to know it. 😆 Hope u understand,.. I'm really not a native speaker. 🤣 

zone.js is cool but I don't miss it in the browser and in Node.js there is the Async hook API and AsyncLocalStorage which is much better IMHO.

---
> Does Type.getTypes() return classes* which are not exported? If so, is ctor just undefined?

It depends.. Regarding to [configuration](https://github.com/Hookyns/ts-reflection/wiki/Configuration) we have two metadata types, `inline` and `typelib`. 

**In `inline` mode, it is quite unreliable**, because metadata (serialized information about types) are per-file. If you use getType<OfSomething>() in file, that file will contain metadata of that type (metadata will be duplicit when getType<OfSomething>() will be used in another file with same type). By importing files/modules, those metadata are appended into global storage. So there may or may not be all the types. It depends if you first import all the modules and then u do something or if you do something inside those modules immediately. 
`ctor` of unexported types will be undefined except cases when you use getType<Type>() inside module scope. Cuz of some changes, there is an [issue](https://github.com/Hookyns/ts-reflection/issues/23).

**`typelib` collects all the metadata into one JS file (auto-imported), so it will return all the types.** But it depends on the configuration. 

Now, only types used in getType<>() will be in metadata (and all the related types used in properties, parameters etc..). It is an optimization cuz of browser usage (it is possible to generate megabytes of metadata but nobody want this in the browser). In that case, Type.getTypes() is useless. But we are working on include/exclude configuration, which will generate metadata of all the types from files/modules matched by those glob patterns. Only then Type.getTypes() will be useful.

**Comment by rezonant** [comment]
> There are a lot of things that people don't know or don't use because they don't know what is it good for. Among other languages, I also work with C# and it has many nice features which nobody miss in JavaScript/Node.js because they don't know that something like that is even possible so they don't know they want to know it. 😆 Hope u understand,.. I'm really not a native speaker. 🤣
> 
> zone.js is cool but I don't miss it in the browser and in Node.js there is the Async hook API and AsyncLocalStorage which is much better IMHO.

When you are building APIs is there a logging library for Node.js that outputs the relevant request ID in all logs that are the result of processing that request? This is standard on Ruby on Rails but there is no way to do it correctly in Node.js that I'm aware of short of (A) processing only one request at a time of (B) passing that request ID (or the request itself) at every function call, even if the function could be used outside of a request.

This is built in to `@alterior/logging`, you don't have to do anything but use the Logger class. And it is ergonomic, you can simply access the current request as `WebEvent.request` globally from anywhere. Even when using async code and multiple requests being processed at the same time, `WebEvent.request` will return the correct request.

When you are testing async code, what happens if you forget an `await` or you forget to call `done()` ? Why can't this work?

```
describe('thing', it => {
    it('works' , () => {
        setTimeout(() => { throw new Error('uh oh, the test failed'); });
    });
});
```

Almost every test framework will report that this test has succeeded, and then you'll get an unhandled promise rejection error later which doesn't fail the test run and thus doesn't fail your build. But using Zone.js this code works, see [razmin](https://github.com/rezonant/razmin) for a library which does this (also mine).

> There are a lot of things that people don't know or don't use because they don't know what is it good for.

I think that is why you don't miss Zone.js :-)

**Comment by rezonant** [comment]
> It depends.. Regarding to [configuration](https://github.com/Hookyns/ts-reflection/wiki/Configuration) we have two metadata types, inline and typelib.

Sorry I think I should clarify what I mean. 

```typescript
class Foo {
}
```

Here `Foo` is not exported, so there's no way to get the constructor. Assuming tst-reflect is configured appropriately, would `Type.getTypes()` return an entry for `Foo`? If so, what would `foosType.ctor()` do?

Also, the return value of `ctor()` seems to differ between CommonJS and ESM modes. `import()` returns a Promise, but `require()` does not. The call signature there seems to indicate that you will get back a Function but you won't if your target is ESM (it would be `Promise<Function>`). In that case it would be required to do `await ctor()` (which indeed would work fine for both ESM and CJS).

**Comment by rezonant** [comment]
Additionally it does occur to me that when tst-reflect is collecting all class metadata into the library, you can indeed get the type for an arbitrary value, but it is an expensive operation indeed:

```typescript
async function getTypeOfValue(value : any) {
  return Type.getTypes().find(x => value.constructor === await x.ctor());
}
```

This would blow up lazy loading on the web unfortunately and could be an expensive call if you have thousands or tens of thousands of types in there. Also unclear if it would return only _my_ compilation unit or if it would include all imported packages which also were compiled with the transformer. Or indeed if when the subpackage iterated over its own types in the context of being included by the parent package, whether it would get its own types or the ones of the parent package that was compiled "last" ?

EDIT: on this last point zone.js could* solve the issue :-) -- Though in thinking closer it _would_ require you to load the entire compilation unit _in_ a zone which could be done but would be a huge lift for the transformer, and not worth it.

**Comment by Hookyns** [comment]
Node.js's Async Hook API is almost the same as zone.js but Async Hook API is native.
There is an issue in zone.js to let zone.js use native Async Hook API when running in Node.js. https://github.com/angular/angular/issues/32796

> I think that is why you don't miss Zone.js :-)

Well I use async context on server for Unit of Works, request context and so on... 
I have no usage for that on client side. BTW I use React and it has own Context API,.. so..

**Comment by rezonant** [comment]
> Now, only types used in getType<>()

Does the class emit when there's a decorator on it? So far that's been the conventional metadata opt-in, should probably do that if nothing else other than backwards compat.

I also had another question, how are you handling emitDecoratorMetadata-- I've found that allowing it to be used while rtti is active actually makes a correct code emission almost impossible because it adds it's own import for it and that cannot be detected until a later compilation stage that transformers don't run in. The solution here is to output that format ourselves if the user has it on, but during the transformation stage we turn it off in TS itself so that it doesn't interfere

EDIT:Actually that night not matter for tst-reflect since it doesn't output this way at all

**Comment by Hookyns** [comment]
I'll respond later.. different time zones.. 😄  going to bed.. I'm super tired.

**Comment by Hookyns** [comment]
Just a quick response to `ctor: Function`.

We create that metadata by transformer in transpile(/compile) time and we generate:
```javascript
{
 n: "NameOfTheType",
 .....
 ctor: () => require("./path/to/that/module.js")
}
```

So in resulted JavaScript (which enters the Webpack, Parcel, ....) there is static path so bundlers know that module and include that module. It is not lazy imported.

**Comment by rezonant** [comment]
> It is not lazy imported.

I don't mean that you are doing lazy importing, though when you are targeting ESM and you cannot use require() (ie when not using a bundler), then it effectively is lazyloading. I believe tst-reflect supports emitting import() when targeting ESM (from a cursory reading), is that not the case?

What I mean is that in the case where the end user is using a lazy loading technique it would cause an import() for every file during this one loop, and thus lazy loading would be rendered ineffective, since we just caused the entire app to be loaded by calling ctor() on all the types

EDIT: I'll put all my other responses here to avoid pinging you over the night. Get some rest! I'm a night owl and in PST in the US so if you want to do a live chat or a voice call or something to discuss, that might be possible :-D

> Node.js's Async Hook API is almost the same as zone.js but Async Hook API is native.
> There is an issue in zone.js to let zone.js use native Async Hook API when running in Node.js. https://github.com/angular/angular/issues/32796

Yes, see also https://github.com/angular/angular/issues/31730, I've been following both issues.
If Zone.js can be mapped onto async_hooks that will be great, or if we can get a general Promise hooks spec available on the browser then fantastic. 

Personally I think Zones should be a native part of Ecmascript though. We have no other way to do thread locals.

> I think that is why you don't miss Zone.js :-)

> Well I use async context on server for Unit of Works, request context and so on... 
> I have no usage for that on client side. BTW I use React and it has own Context API,.. so..

Does React's context API patch the world like Zone.js though? Seems like most of what I've put above would not be possible or at least would only be possible in a library tightly coupled to React. I use Angular for UI, how do we share code? Shouldn't this work regardless of UI framework?

**Comment by capaj** [comment]
I will most likely use both of these libs on the new major of decapi.
Classes are runtime values and so you should ideally be able to do reflection on them at runtime.
Interfaces and types on the other hand need the tst-reflect approach.
I mostly need functionality of rtti, but when modelling Graphql interfaces from typescript interfaces tst-reflect will come in and save the day. All I need is for users of decapi to pass the interface as a generic argument.

So if you are thinking about joining forces, the result would ideally be a merge of both current approaches as I imagine many other consumers will want both too. For example https://github.com/MichalLytek/type-graphql

**Comment by rezonant** [comment]
> Interfaces and types on the other hand need the tst-reflect approach.

This is changing with https://github.com/rezonant/typescript-rtti/pull/7

**Comment by Hookyns** [comment]
Okay guys, I've just woke up and I figured it out. https://github.com/Hookyns/ts-reflection/issues/25

**Comment by rezonant** [comment]
@capaj v0.0.20 is released, supports interfaces via `reflect<Interface>()` and obtaining interface tokens via `reify<Interface>()`.

**Comment by rezonant** [comment]
So I think I want to take `reflect<T>()` out of this project and into its own. The promise of `typescript-rtti` is that you need only interact with the API, and the reflect-metadata format underneath it is an implementation detail. Indeed, `reflect`/`ReflectedClass` already supports classes annotated with `emitDecoratorMetadata` and I think it should be made to support any other competing or future metadata format that arises so that users need not care which transformer solution was used for which library that they depend on, for instance.

**Comment by rezonant** [comment]
@Hookyns Why did you choose not to use `reflect-metadata`?

**Comment by Hookyns** [comment]
@capaj https://github.com/Hookyns/ts-reflection/issues/25 is implemented and will be released tomorrow. 🚀 

----

> Why did you choose not to use reflect-metadata?

`tst-reflect` is beyond its scope. Using `reflect-metadata` would be overwhelming with almost no gain. IMHO. 
You have to decorate everything and you cannot get type of types from 3rd party modules, when they are not decorated.. 
And well,.. we have no issue with circular references.

**Comment by rezonant** [comment]
I think there's some confusion, reflect-metadata is simply a (defacto) standard way to attach metadata to objects. It doesn't have anything to do with decorators itself, though it is the way that Typescript enables emitDecoratorMetadata.

> you cannot get type of types from 3rd party modules

I don't understand this, emitDecoratorMetadata works across modules and so do other solutions that use reflect-metadata like typescript-rtti. 

> well,.. we have no issue with circular references.

reflect-metadata also works fine with circular references. typescript-rtti fully supports circular and forward referencing because we use the "forward reference" pattern. This means we always store a function that resolves to the type metadata so that the name binding is late.

I ask this because reflect-metadata is a great way to ensure interoperability. Your solution is now using private symbols which would require another library to depend on yours to provide interoperability, whereas using reflect-metadata as the underlying storage system would not, while still having the benefits of using symbols to occlude the metadata (underneath the covers that is exactly what reflect-metadata uses).

**Comment by rezonant** [comment]
Take for instance this example from our README:

```typescript
```typescript

//input 

export class B {
    constructor(
        readonly a : A
    ) {
    }
}

// output

const __RtΦ = (k, v) => Reflect.metadata(k, v);
//...
B = __decorate([
    __RtΦ("rt:P", ["a"]),
    __RtΦ("rt:p", [{ n: "a", t: () => A, f: "R" }]),
    __RtΦ("rt:f", "C$")
], B);
```

This call declares a reflect-metadata key of "rt:p" on the class B:

```typescript
__RtΦ("rt:p", [{ n: "a", t: () => A, f: "R" }]),
```

I could access this metadata from any other module (across packages) as `Reflect.getMetadata("rt:p", B)`. That would yield the array shown above. 

This key:

```typescript
t: () => A
```

...is a "type resolver". If we had emitted `t: A`, it would not work with circular or forward references. But since it is wrapped in a function, we simply need to evaluate the function first, and we will get back `A`, even if there is a circular or forward reference situation. This is because at the time that you evaluate the function, all of the initialization for the symbols has completed. 

This is one of the important improvements that typescript-rtti has over the default emitDecoratorMetadata scheme- it's one of the changes that Typescript themselves have wished they had done, because we never would have had issues with circular references in the first place.

**Comment by rezonant** [comment]
Still curious about the above. 

In other news:
- typescript-rtti now also supports reflection on function declarations, expressions and arrow functions which brings RTTI to function-based paradigms. 
- the test suite is upgraded- we now build several larger codebases and run their tests under the transformer to ensure correct builds

I need to dogfood this within my bigger private codebases before I think a 1.0 is warranted, but we're well on the way here.

**Comment by Hookyns** [comment]
I'm gonna answer in few days. I'm super bussy cuz of work. 😞

**Comment by rezonant** [comment]
In 0.2.0 we sort of have metadata libraries now, but its per-file. This is to support emitting recursive types. Looking forward to continuing the conversation!

**Comment by rezonant** [comment]
You handled generic transients beautifully with the arguments patching :-D
```typescript
export function cool<T>(...args) {
    console.dir(args);
    return getType<T>();
}
```

```javascript
function cool(...args) {
    var __genericParams__ = args.splice(args.length - 1, 1)[0];
    delete arguments[--arguments.length];
    console.dir(args);
    return (__genericParams__ && __genericParams__.T);
}
```

I think this is the only way to do it.
However, this does not currently seem to work across separately built projects. Indeed, it's not clear how tst-reflect would know whether the API is used in a function that is only represented as a FunctionSignature in the d.ts. I get the following when I try it:

```
> ttsc -b

D:\Dev\rezonant\tst-reflect-test\node_modules\typescript\lib\typescript.js:114725
                throw e;
                ^

Error: Debug Failure. False expression: Expected fileName to be present in command line
    at Object.getOutputFileNames (D:\Dev\rezonant\tst-reflect-test\node_modules\typescript\lib\typescript.js:107595:18)
    at getOutPathForSourceFile (D:\Dev\rezonant\tst-reflect-test\node_modules\tst-reflect-transformer\dist\helpers.js:257:19)
    at getExportOfConstructor (D:\Dev\rezonant\tst-reflect-test\node_modules\tst-reflect-transformer\dist\getExports.js:28:56)
    at getTypeDescription (D:\Dev\rezonant\tst-reflect-test\node_modules\tst-reflect-transformer\dist\getTypeDescription.js:242:79)
    at getTypeCall (D:\Dev\rezonant\tst-reflect-test\node_modules\tst-reflect-transformer\dist\getTypeCall.js:20:71)
    at processGetTypeCallExpression (D:\Dev\rezonant\tst-reflect-test\node_modules\tst-reflect-transformer\dist\processGetTypeCallExpression.js:21:46)
    at mainVisitor (D:\Dev\rezonant\tst-reflect-test\node_modules\tst-reflect-transformer\dist\visitors\mainVisitor.js:26:93)
    at Context._visitor (D:\Dev\rezonant\tst-reflect-test\node_modules\tst-reflect-transformer\dist\contexts\Context.js:9:35)
    at visitNodes (D:\Dev\rezonant\tst-reflect-test\node_modules\typescript\lib\typescript.js:87058:48)
    at Object.visitEachChild (D:\Dev\rezonant\tst-reflect-test\node_modules\typescript\lib\typescript.js:87385:173)
```

Project is at https://github.com/rezonant/tst-reflect-test

You need to `npm link` these two together:

```
cd project2
npm link
cd ..
cd project1
npm link test-project-2
cd ..
npm run build
```

I want to add this sort of functionality myself, but I don't know how we can solve this issue.

**Reply by Hookyns** [reply]
So you don't want to merge our projects and join our effort, but you want to steal all the `tst-reflect` features you never thought about. Excellent! I think, our discussion ends here. 👎 I hope you have at least a little dignity and you'll link the source; it's not your idea.

**Reply by rezonant** [reply]
@Hookyns I'm not out to "steal features" from tst-reflect. I'm out to produce a conceptually sound solution to the problem of runtime reflection. Your solution with tst-reflect is very nice, but there are parts of it that bother me, most notably the lack of integration with elements like reflect-metadata that are the standard and expected way to do it. 

typescript-rtti has all kinds of features that tst-reflect does not have, and you are more than welcome, and in fact _encouraged_ to implement those features. 

I missed this message initially and just came to explain how I solved the opt-in problem because I want you to benefit from it too so that tst-reflect can do this cross-package. 

PS: Besides, the feature I'm talking about _doesn't even work correctly in tst-reflect_. I shared this with you here instead of on your project so you could fix it without slowing down adoption or making people look for alternatives, which wouldn't have been nice of me either, would it. I didn't have to do any of that. I thought we could work together to find a solution for both our projects.

I don't want to join our projects _right now_ at least in the way you would want me to- I want to collaborate on a shared reflection API as I've mentioned prior. 

PPS (sorry, I'm a bit upset about this so I'm ranting): 
> So you don't want to merge our projects and join our effort

Have you considered that you could also join _this_ project? Never once did you seem to consider that in our discussions. Seems like you assumed this project was far behind yours but I'm not sure that was ever the case and I don't think it's the case now. Regardless, you are always welcome.

PPPS:

I only investigated implementing this as ahupp requested it here in this discussion space: https://github.com/typescript-rtti/typescript-rtti/discussions/3 -- later I realized that online there are a lot of people who are looking for a way to do "generic passing" in the wild, and since all the rest of the functionality is done it made sense to try to find a solution. I tried the feature out in tst-reflect, curious how it could handle the opt-in problem to avoid performance problems and was impressed, but it was trivially easy to break it.

**Comment by rezonant** [comment]
@Hookyns,

Putting aside your disagreement about me implementing the ability to pass generics, I wanted to share what typescript-rtti@0.4.0 does and how it might help you solve the cross-package issue.

I looked at a _ton_ of ways to solve this issue, with the final way I thought it could work being passing the opt-in information in the JSDoc block, since we can modify the `.d.ts` file to include new comments. However, the `afterDeclarations` transformer path has no access to the function body, meaning you cannot detect whether the function uses the reflection API in order to _add_ the optin. We could make users put it on there manually, but I realized there was a better way!

The new feature is called "call site reflection" and it can do more than just generic parameters-- it can also reflect on the types of arguments at the call-site. For instance, you could do this:

```typescript
import { reflect, CallSite } from 'typescript-rtti';

export function typeOf(value : any, callSite? : CallSite): ReflectedTypeRef {
    return reflect(callSite).parameters[0];
}

let a : string | number = 123;
expect(typeOf(a).is('union')).to.be.true;
expect(typeOf(a).as('union').types[0].isClass(String)).to.be.true;
```

And of course the regular generic thing:

```typescript
import { reflect, CallSite } from 'typescript-rtti';

export function matchesShape<T>(value : any, callSite? : CallSite): ReflectedTypeRef {
    return reflect<T>().matchesShape(value);
}
```

As you can see, this is pretty different than what tst-reflect is doing, and I think it adds some more possibilities. I _encourage_ you to consider how this works and feel free to "steal" anything you want from it. It's all MIT.

**Reply by Hookyns** [reply]
Interesting idea. 👍 
I try to avoid the programmer having to modify the code cuz of reflection, but there is no other way to achieve such behavior. 

PS: You probably doesn't read the [docs](https://github.com/Hookyns/tst-reflect/wiki/Custom-Generic-Methods-&-Functions). There is a `@reflect` JSDoc tag and decorator which "marks" methods and functions. It works cross-package and with polymorphism.

**Reply by rezonant** [reply]
Nice! I've thought about offering this kind of optin as well because spread args cannot work with the callSite optin, but it is not there right now.

**Reply by rezonant** [reply]
But you probably should fix that crash.

**Comment by rezonant** [comment]
If you want to collaborate, there are two things I'm interested in.
1. I want to collaborate on a written specification for how metadata should be encoded by transformers. This should be the base ground rules and be extensible to allow for innovation.
2. I want to collaborate on a shared reflection library (not a transformer) that supports the metadata emitted by emitDecoratorMetadata, typescript-rtti _and_ tst-reflect. I want that library to also implement the standard specification we come up with. 

There's no point pretending either one of our projects are anywhere worthy of being some kind of official solution, and we should collaborate and compete on those grounds. The innovation gained will be better than the feelings lost.

**Comment by levrik** [comment]
@rezonant @Hookyns
Just FYI on other approaches to TypeScript reflection you should also have a look at Deepkit. While being a complete framework it is also powered by some custom TypeScript reflection engine.
https://deepkit.io/library/type
https://github.com/deepkit/deepkit-framework/tree/master/packages/type
https://deepkit.io/documentation/type/reflection
The package actually handles way more than just type reflection (serialization, deserialization, etc) but the approach taken here is very interesting. The whole concept is described in detail here: https://github.com/microsoft/TypeScript/issues/47658
